core/common/src/main/java/alluxio/conf/PropertyKey.java

```diff
@@ -2657,6 +2657,18 @@ public String toString() {
           .setDefaultValue("1hour")
           .setScope(Scope.WORKER)
           .build();
+  public static final PropertyKey WORKER_MASTER_PERIODICAL_RPC_TIMEOUT =
+      new Builder(Name.WORKER_MASTER_PERIODICAL_RPC_TIMEOUT)
+          .setDefaultValue("5min")
```

Comment:
Does this control the connection phase, or the entire RPC duration as well?

Also, is 5m too long? This means if there is a failover, workers will take 5 minutes to re-connect?

How did failovers work before this parameter?

Reply:
This only controls the RPC duration of selected periodical worker connections. Control the entire RPC duration.
The worker hangs problem only exists if the previous leading master is network partitioned (not responding, instead of rejection). 
I give a long timeout in the assumption that this case rarely happens and try not to affect our normal RPCs. A frequently called RPC is getPinList(). If the cluster has a large amount of files/dirs get pinned, i worried that if the timeout is short, the getPinList() may keep getting timeout unexpectedly.

Reply:
Before this parameter, if the previous leading master is network partitioned (not responding), our workers will hang in periodical RPCs and fail to response to any requests or register with the new leading master. After the previous leading master come back, workers will receive exception message from previous leading master and continue functioning (register with new leading master)

When network condition is good, the rpcs should direct fail when connecting to previous leading master and no hanging should happen.

Reply:
@bf8086 What's the timeout that you think is reasonable. I saw we getPinList every second, probably 30sec or 1min timeout is good for this case?

Reply:
@LuQQiu 
I see. Let's say we have a 1m timeout, and the worker heartbeat takes longer than 1m, even though nothing has happened to the master (no failover, no network flakiness). What will happen to the worker? Will it just try to reconnect again?

If the worker heartbeat always takes more than 1m, does that mean that worker will never be able to successfully perform a heartbeat?

Reply:
Yeah for both questions, it will create a new client and retry operations. If the operations consistent taking longer than 1m, then heartbeat will never succeed. That's why i prefer to give the timeout a larger value.
This problem exists in previous versions but at least never be caught by us. It could be the network partition happens rarely and can be recovered in a short period of time. This change mainly protects the RPC from hanging forever.

Reply:
@LuQQiu I see. Then, maybe we do want to make the timeout longer.

From an earlier comment, you said typically, when a master does a failover, the worker will immediately know, because the old master would actively reject the connection? Is this correct? That means this timeout will not affect normal failovers, but only unreachable old masters? If this is the case, then this is more reason to make the timeout long.

Thanks for the explanations.

Reply:
Yeah, normally the worker will immediately know it cannot connect to the old leading master and connect to the new leading master. This timeout will not affect normal failover and only be useful when old leading master is unreachable.